creator-node/src/fileManager.js

```diff
@@ -97,6 +97,217 @@ async function copyMultihashToFs(multihash, srcPath, logContext) {
+    if (!found) {
+      throw new Error(`Did not find multihash=${multihash} from network`)
+    }
+
+    decisionTree.recordStage({
+      name: 'Found file from network'
+    })
+
+    return
+  } catch (e) {
+    decisionTree.recordStage({
+      name: `Failed to find file from network`,
+      data: { errorMsg: e.message }
+    })
+  }
+
+  // error if file was not found on any gateway
+  const errorMsg = `Failed to retrieve file for multihash ${multihash} after trying entire network`
+  decisionTree.recordStage({
+    name: errorMsg
+  })
+  throw new Error(errorMsg)
+}
+
+/**
+ * Fetches a multihash via the /ipfs route, writes to disk, and verifies that the CID is what we
+ * expect it to be with retries
+ * @param {Object} param
+ * @param {function} param.bail the npm module async-retry bail fn
+ * @param {string} param.contentUrl the target content node ipfs gateway route
+ * @param {number} param.trackId the track id if one is associated with the multihash fetched
+ * @param {Object} param.decisionTree an instance of DecisionTree
+ * @param {string} param.path the path at which to save the fetched multihash from
+ * @param {Object} param.logger
+ */
+async function fetchFileFromTargetGatewayAndWriteToDisk({
+  bail,
+  contentUrl,
+  trackId,
+  multihash,
+  decisionTree,
+  path,
+  logger
+}) {
+  let response
+  try {
+    const fetchReqParams = {
+      method: 'get',
+      url: contentUrl,
+      responseType: 'stream',
+      timeout: 20000 /* 20 sec - higher timeout to allow enough time to fetch copy320 */
+    }
+
+    if (trackId) {
+      fetchReqParams.params = {
+        trackId
+      }
+    }
+
+    response = await axios(fetchReqParams)
+  } catch (e) {
+    if (
+      e.response?.status === 403 || // delist
+      e.response?.status === 401 || // unauth
+      e.response?.status === 400 // bad request
+    ) {
+      bail(
+        new Error(
+          `Content is delisted, request is unauthorized, or the request is bad with statusCode=${e.response?.status}`
+        )
+      )
+      return
+    }
+
+    throw new Error(
+      `Failed to fetch content=${multihash} with statusCode=${e.response?.status}. Retrying..`
+    )
+  }
+
+  if (!response || !response.data) {
+    throw new Error(`Received empty response`)
+  }
+
+  decisionTree.recordStage({
+    name: 'Retrieved file from target gateway',
+    data: { url: contentUrl }
+  })
+
+  await Utils.writeStreamToFileSystem(response.data, path)
+
+  decisionTree.recordStage({
+    name: 'Wrote file to file system after fetching from target gateway',
+    data: { expectedStoragePath: path }
+  })
+
+  const CIDMatchesExpected = await Utils.verifyCIDMatchesExpected({
+    cid: multihash,
+    path: path,
+    logger
+  })
+
+  if (!CIDMatchesExpected) {
+    try {
+      await fs.unlink(path)
+    } catch (e) {
+      logger.error(`Could not remove file at path=${path}`)
+    }
+    throw new Error('CID does not match what is expected to be')
+  }
+
+  logger.info(
```

Comment:
nit - these internal stages shouldn't be emitting logs, instead just recording a decisionTree stage, and the full tree is logged once

Reply:
that's fair. i'll remove these logs and add to the decision tree instead